vm/systemSmartContracts/staking.go

```diff
@@ -1406,13 +1413,134 @@ func (r *stakingSC) getOwner(args *vmcommon.ContractCallInput) vmcommon.ReturnCo
+	}
+
+	numNodesToStake := big.NewInt(0).SetBytes(args.Arguments[0]).Uint64()
+	waitingListData, err := r.getFirstElementsFromWaitingList(uint32(numNodesToStake))
+	if err != nil {
+		r.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+	if len(waitingListData.blsKeys) == 0 {
+		r.eei.AddReturnMessage("no nodes in queue")
+		return vmcommon.Ok
+	}
+
+	for i, blsKey := range waitingListData.blsKeys {
+		stakedData := waitingListData.stakedDataList[i]
+		r.activeStakingFor(stakedData)
+		err = r.saveStakingData(blsKey, stakedData)
+		if err != nil {
+			r.eei.AddReturnMessage(err.Error())
+			return vmcommon.UserError
+		}
+
+		// remove from waiting list
+		inWaitingListKey := r.createWaitingListKey(blsKey)
+		r.eei.SetStorage(inWaitingListKey, nil)
+
+		// return the change key
+		r.eei.Finish(blsKey)
+		r.eei.Finish(stakedData.RewardAddress)
+	}
+
+	err = r.updateWaitingListToNewFirstKey(waitingListData)
+	if err != nil {
+		r.eei.AddReturnMessage(err.Error())
+		return vmcommon.UserError
+	}
+
+	return vmcommon.Ok
+}
+
+func (r *stakingSC) updateWaitingListToNewFirstKey(waitingListData *waitingListReturnData) error {
+	waitingListHead, err := r.getWaitingListHead()
+	if err != nil {
+		return err
+	}
+	if waitingListHead.Length == 0 {
+		return nil
+	}
+	if len(waitingListData.lastKey) == 0 {
+		r.eei.SetStorage([]byte(waitingListHeadKey), nil)
+		return nil
+	}
+	if waitingListData.afterLastjailed {
+		waitingListHead.LastJailedKey = nil
+	}
+
+	waitingListHead.Length = waitingListHead.Length - uint32(len(waitingListData.blsKeys))
+	waitingListHead.FirstKey = waitingListData.lastKey
+	err = r.saveWaitingListHead(waitingListHead)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *stakingSC) getFirstElementsFromWaitingList(numNodes uint32) (*waitingListReturnData, error) {
+	waitingListData := &waitingListReturnData{}
+
+	waitingListHead, err := r.getWaitingListHead()
+	if err != nil {
+		return nil, err
+	}
+	if waitingListHead.Length == 0 {
+		return waitingListData, nil
+	}
+
+	blsKeysToStake := make([][]byte, 0)
+	stakedDataList := make([]*StakedDataV2_0, 0)
+	index := uint32(1)
+	nextKey := make([]byte, len(waitingListHead.FirstKey))
+	copy(nextKey, waitingListHead.FirstKey)
+	for len(nextKey) != 0 && index <= waitingListHead.Length && index <= numNodes {
+		element, errGet := r.getWaitingListElement(nextKey)
+		if errGet != nil {
+			return nil, errGet
+		}
+
+		if bytes.Equal(nextKey, waitingListHead.LastJailedKey) {
+			waitingListData.afterLastjailed = true
+		}
+
+		stakedData, errGet := r.getOrCreateRegisteredData(element.BLSPublicKey)
+		if errGet != nil {
+			return nil, errGet
+		}
+
+		blsKeysToStake = append(blsKeysToStake, element.BLSPublicKey)
+		stakedDataList = append(stakedDataList, stakedData)
+		index++
+		copy(nextKey, element.NextKey)
+	}
+
+	waitingListData.blsKeys = blsKeysToStake
+	waitingListData.stakedDataList = stakedDataList
+	waitingListData.lastKey = nextKey
```

Comment:
So, actually lastKey will remain the first element in waiting list ?

Reply:
Yes. All the rest is deleted. As all before when staked. Moved from waiting list to staking